packages/runtime/container-runtime/src/runningSummarizer.ts

```diff
@@ -329,18 +350,77 @@ export class RunningSummarizer implements IDisposable {
             // Another summary is currently being generated.
             return { alreadyRunning: this.generator.waitSummarizing() };
         }
-        const result = this.generator.summarize(`onDemand;${reason}` as `onDemand;${string}`, options);
-        result.receivedSummaryAckOrNack.finally(() => this.checkRerunHeuristics());
+        const result = this.generator.summarize(onDemandReason, options);
+        result.receivedSummaryAckOrNack.finally(() => this.checkSummarizeAgain());
         return result;
     }
 
-    /** After summarizing, we should rerun the heuristics to see if we already need to summarize again. */
-    private checkRerunHeuristics() {
-        if (this.tryWhileSummarizing) {
+    public enqueueSummarize({
+        reason,
+        afterSequenceNumber = 0,
+        ...options
+    }: IEnqueueSummarizeOptions): ISummarizeResults {
+        const onDemandReason = `enqueue;${reason}` as const;
+        // If afterSequenceNumber is provided, the on-demand summary is "enqueued" to run
+        // once an eligible op comes in with sequenceNumber >= afterSequenceNumber.
+        if (this.enqueuedSummary !== undefined) {
+            this.enqueuedSummary.reason = onDemandReason;
```

Comment:
Based on our discussion, this might not be the right approach if there is already an enqueued summary. We should return enough information in this scenario for the caller to decide if the currently enqueued summary satisfies their need or they can enqueue another one once this one completes.

Reply:
Yeah I was thinking that at first, but then I realized, I think it might be simpler to just overwrite. If you look at the other place, I switched the order so it sets `this.enqueuedSummary` to undefined _before_ attempting to summarize, so we prevent the case I was worried about in my old comment.

Reply:
The case I am thinking about is when we user has enqueued a summary for say `seq# 10` and `fullTree: true` and the summary hasn't started. Then another summary is enqueued for `seq# 20` and `fullTree: false`. Now, the first caller will get result for `seq# 20` which is fine but the results will not have full tree.
Also, if the call order is reversed, then the call site waiting for `seq# 20` will get result for `seq# 10`.

Reply:
Oh that's right. I forgot we had talked about that.

Reply:
Okay I've updated it to include an optional override parameter. If override is passed, then it will abort the existing enqueued summarize attempt, failing it, and then replacing it with this attempt. If not, then it will reject this attempt.